File: javascript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
  constructor(data) {
    this.data = data
    this.left = null
    this.right = null
  }

  insert(value, root = this) {
    if (root.data >= value) {
      if (root.left == null) {
        root.left = new BinarySearchTree(value)
        return
      }
      this.insert(value, root.left)
    }
    else if (root.data < value) {
      if (root.right == null) {
        root.right = new BinarySearchTree(value)
        return
      }
      this.insert(value, root.right)
    }
  }

  each(func, root = this) {
    if (root.left) {
      this.each(func, root.left)
    }
    func(root.data)
    if (root.right) {
      this.each(func, root.right)
    }
  }
}

File: kotlin/binary-search-tree/src/main/kotlin/BinarySearchTree.kt
class BinarySearchTree<T : Comparable<T>> {

    data class Node<T>(
        val data: T,
        var left: Node<T>?,
        var right: Node<T>?
    )

    var root: Node<T>? = null

    fun insert(value: T) {
        if (root == null) {
            root = Node(value, null, null)
        } else {
            insert(value, root!!)
        }
    }

    private fun insert(value: T, root: Node<T>) {
        if (value <= root.data) {
            if (root.left == null) {
                root.left = Node(value, null, null)
             } else {
                insert(value, root.left!!)
            }
        } else {
            if (root.right == null) {
                root.right = Node(value, null, null)
            } else {
                insert(value, root.right!!)
            }
        }
    }

    fun asSortedList(): List<T> {
        val list = mutableListOf<T>()
        asSortedList(root, list)
        return list
    }

    fun asSortedList(root: Node<T>?, list: MutableList<T>) {
        if (root == null) {
            return
        }
        asSortedList(root.left, list)
        list.add(root.data)
        asSortedList(root.right, list)
    }

    fun asLevelOrderList(): List<T> {
        val output = mutableListOf<T>()

        val queue = mutableListOf<Node<T>?>()
        queue.add(root)

        while(queue.isNotEmpty()) {
            val node = queue.removeAt(0) ?: continue
            output.add(node.data)
            queue.add(node.left)
            queue.add(node.right)
        }

        return output
    }
}
